File: Python/baekjoon/silver/s3/달팽이.py
n = int(input())
find_number = int(input())

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

direction = 0
x, y = 0, 0
cnt = n * n

map = [[0] * n for _ in range(n)]

ans_x, ans_y = 0, 0

while cnt > 0:
    map[x][y] = cnt
    cnt -= 1

    nx = x + dx[direction]
    ny = y + dy[direction]

    if 0 <= nx < n and 0 <= ny < n and map[nx][ny] == 0:
        x = nx
        y = ny
    else:
        direction = (direction + 1) % 4
        x = x + dx[direction]
        y = y + dy[direction]

    if find_number == cnt:
        ans_x, ans_y = x, y

for i in range(n):
    for j in range(n):
        print(map[i][j], end=' ')
    print()

print(ans_x + 1, ans_y + 1)

File: kotlin_java/src/baekjoon/silver/s3/달팽이.kt
package baekjoon.silver.s3


fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val target = readLine().toInt()
    val map = Array(n) { Array(n) { 0 } }
    val dx = intArrayOf(1, 0, -1, 0)
    val dy = intArrayOf(0, 1, 0, -1)
    var dir = 0
    var (ansX, ansY) = listOf(0, 0)

    var cnt = n * n
    var (x, y) = listOf(0, 0)

    while (cnt > 0) {
        map[x][y] = cnt
        cnt--

        val nx = x + dx[dir]
        val ny = y + dy[dir]

        if (nx in 0..<n && ny in 0..<n && map[nx][ny] == 0) {
            x = nx
            y = ny
        } else {
            dir = (dir + 1) % 4
            x += dx[dir]
            y += dy[dir]
        }

        if (target == cnt) {
            ansX = x
            ansY = y
        }
    }

    val sb = StringBuilder()
    repeat(n) { i ->
        repeat(n) { j ->
            sb.append(map[i][j]).append(" ")
        }
        sb.appendLine()
    }
    print(sb)
    print("${ansX + 1} ${ansY + 1}")
}